src/dispdrivers/vgacontroller.cpp

```diff
@@ -151,33 +151,95 @@ void IRAM_ATTR VGAController::VSyncInterrupt(void * arg)
+    case PaintMode::AND:
+    case PaintMode::ANDNOT:
+      return [&] (int X, int Y, uint8_t pattern) { VGA_PIXEL(X, Y) &= pattern; };
+    case PaintMode::XOR:
+      return [&] (int X, int Y, uint8_t pattern) { VGA_PIXEL(X, Y) ^= pattern; };
+    case PaintMode::Invert:
+      return [&] (int X, int Y, uint8_t pattern) { VGA_INVERT_PIXEL(X, Y); };
+    default:  // PaintMode::NoOp
+      return [&] (int X, int Y, uint8_t pattern) { return; };
+  }
+}
+
+std::function<void(int Y, int X1, int X2, uint8_t pattern)> VGAController::fillRowLamda(PaintMode mode)
+{
+  switch (mode) {
+    case PaintMode::Set:
+      return [&] (int Y, int X1, int X2, uint8_t pattern) { rawFillRow(Y, X1, X2, pattern); };
+    case PaintMode::OR:
+    case PaintMode::ORNOT:
+      return [&] (int Y, int X1, int X2, uint8_t pattern) { rawORRow(Y, X1, X2, pattern); };
+    case PaintMode::AND:
+    case PaintMode::ANDNOT:
+      return [&] (int Y, int X1, int X2, uint8_t pattern) { rawANDRow(Y, X1, X2, pattern); };
+    case PaintMode::XOR:
+      return [&] (int Y, int X1, int X2, uint8_t pattern) { rawXORRow(Y, X1, X2, pattern); };
+    case PaintMode::Invert:
+      return [&] (int Y, int X1, int X2, uint8_t pattern) { rawInvertRow(Y, X1, X2); };
+    default:  // PaintMode::NoOp
+      return [&] (int Y, int X1, int X2, uint8_t pattern) { return; };
+  }
+}
+
+
+
 void IRAM_ATTR VGAController::setPixelAt(PixelDesc const & pixelDesc, Rect & updateRect)
 {
-  genericSetPixelAt(pixelDesc, updateRect,
-                    [&] (RGB888 const & color)          { return preparePixel(color); },
-                    [&] (int X, int Y, uint8_t pattern) { VGA_PIXEL(X, Y) = pattern; }
-                   );
+  auto paintMode = paintState().paintOptions.mode;
+  genericSetPixelAt(pixelDesc, updateRect, getPixelLamda(paintMode), setPixelLamda(paintMode));
 }
 
 
+
 // coordinates are absolute values (not relative to origin)
 // line clipped on current absolute clipping rectangle
 void IRAM_ATTR VGAController::absDrawLine(int X1, int Y1, int X2, int Y2, RGB888 color)
 {
+  auto paintMode = paintState().paintOptions.mode;
+
   genericAbsDrawLine(X1, Y1, X2, Y2, color,
-                     [&] (RGB888 const & color)                   { return preparePixel(color); },
-                     [&] (int Y, int X1, int X2, uint8_t pattern) { rawFillRow(Y, X1, X2, pattern); },
-                     [&] (int Y, int X1, int X2)                  { rawInvertRow(Y, X1, X2); },
-                     [&] (int X, int Y, uint8_t pattern)          { VGA_PIXEL(X, Y) = pattern; },
-                     [&] (int X, int Y)                           { VGA_INVERT_PIXEL(X, Y); }
+                     getPixelLamda(paintMode),
+                     fillRowLamda(paintMode),
+                     [&] (int Y, int X1, int X2) { rawInvertRow(Y, X1, X2); },
+                     setPixelLamda(paintMode),
+                     [&] (int X, int Y)          { VGA_INVERT_PIXEL(X, Y); }
```

Comment:
these "invert" arguments are redundant and will be removed

(they work with paintOptions.NOT, which the new "invert" mode essentially replaces)

Reply:
they're gone now